flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/metrics/AbstractMetricsHandler.java

```diff
@@ -0,0 +1,141 @@
+ * {@code /metrics?get=X,Y}
+ * The handler will then return a list containing the values of the requested metrics.
+ * {@code [ { "id" : "X", "value" : "S" }, { "id" : "Y", "value" : "T" } ] }
+ *
+ * @param <M> Type of the concrete MessageParameters
+ */
+public abstract class AbstractMetricsHandler<M extends MessageParameters> extends
+	AbstractRestHandler<DispatcherGateway, EmptyRequestBody, MetricCollectionResponseBody, M> {
+
+	private final MetricFetcher metricFetcher;
+
+	public AbstractMetricsHandler(
+			CompletableFuture<String> localRestAddress,
+			GatewayRetriever<DispatcherGateway> leaderRetriever,
+			Time timeout,
+			Map<String, String> headers,
+			MessageHeaders<EmptyRequestBody, MetricCollectionResponseBody, M> messageHeaders,
+			MetricFetcher metricFetcher) {
+		super(localRestAddress, leaderRetriever, timeout, headers, messageHeaders);
+		this.metricFetcher = requireNonNull(metricFetcher, "metricFetcher must not be null");
+	}
+
+	@Override
+	protected CompletableFuture<MetricCollectionResponseBody> handleRequest(
+			@Nonnull HandlerRequest<EmptyRequestBody, M> request,
+			@Nonnull DispatcherGateway gateway) throws RestHandlerException {
+		metricFetcher.update();
+
+		final MetricStore.ComponentMetricStore componentMetricStore = getComponentMetricStore(
+			request,
+			metricFetcher.getMetricStore());
+
+		if (componentMetricStore == null || componentMetricStore.metrics == null) {
+			return CompletableFuture.completedFuture(
+				new MetricCollectionResponseBody(Collections.emptyList()));
+		}
+
+		final Set<String> requestedMetrics = new HashSet<>(request.getQueryParameter(
+			MetricsFilterParameter.class));
+
+		if (requestedMetrics.isEmpty()) {
+			return CompletableFuture.completedFuture(
+				new MetricCollectionResponseBody(getAvailableMetrics(componentMetricStore)));
+		} else {
+			final List<Metric> metrics = getRequestedMetrics(componentMetricStore, requestedMetrics);
+			return CompletableFuture.completedFuture(new MetricCollectionResponseBody(metrics));
+		}
+	}
+
+	/**
+	 * Returns the {@link MetricStore.ComponentMetricStore} that should be queried for metrics.
+	 */
+	@Nullable
+	protected abstract MetricStore.ComponentMetricStore getComponentMetricStore(
+		HandlerRequest<EmptyRequestBody, M> request,
+		MetricStore metricStore);
+
+	private static List<Metric> getAvailableMetrics(MetricStore.ComponentMetricStore componentMetricStore) {
+		return componentMetricStore.metrics
+			.keySet()
+			.stream()
+			.map(Metric::new)
+			.collect(Collectors.toList());
+	}
+
+	private static List<Metric> getRequestedMetrics(
+			MetricStore.ComponentMetricStore componentMetricStore,
+			Set<String> requestedMetrics) throws RestHandlerException {
+
+		final List<Metric> metrics = new ArrayList<>(requestedMetrics.size());
+		for (final String requestedMetric : requestedMetrics) {
+			final String value = componentMetricStore.getMetric(requestedMetric, null);
```

Comment:
If `requestedMetric` is an empty string, we will return HTTP code 404. This is different compared to the legacy handler. If this is a problem, I can change it to the old behavior.

Reply:
I think we actually have a slightly inconsistent behaviour here. In case that `componentMetricStore` or `componentMetricStore.metrics` is `null` we return an empty metrics object. But in case that a requested metric is not there (might not be registered or transferred yet) we fail with a 404. With the current way we handle a non-existent metric store, it would be more consistent to simply not report metrics which are not there.

Furthermore, it could also be the case that only a subset of the requested metrics is not yet registered.

Reply:
Changed the behavior:
* Unknown metrics are not included.
* If all metrics are unknown, an empty list is returned.